GoogleForCreators/web-stories-wp
# Dashboard: PreviewScreen UI Component

## Summary
Sets up a template preview screen.

## Relevant Technical Choices
- Reuses the modal as the overlay so we can get onClose and a11y out of the box and keep the preview in the dashboard and not open a new window. 
- When a template card is clicked or entered down on the associated template is set as the `previewVisible.value`. When that value is present the `PreviewStoryView` is triggered. 
- that `previewVisible.value` is passed to the `PreviewStoryView` as the `story` (since we are technically previewing a template as a story) along with a boolean flag to signal that the story is actually a template and an `onClose` action. 
- When `PreviewStoryView` is rendered the following occurs in order: 
  - In first `useEffect` hook we check that a story is actually present. If there isn't one, there's no need to try and render anything other than an error and a close button(s). Assuming there is a story (there shouldn't be a case where there isn't a story, more just a safe guard against bad ui) then we check to see if the story passed in is a template with the `isTemplate` boolean. Since this PR/feature is for template previews this is the flow that is addressed here. By default, the existing story should be turned into markup without needing to manipulate the template object into a story object. 
  - Following the path of the template preview, we call `createStoryPreviewFromTemplate` from the above useEffect hook. 
  - `createStoryPreviewFromTemplate`  is an story action in our ApiContext that is new and responsible for taking the template object and turning it into a story that can be previewed. Inside of it we call edit-story's `getStoryPropsToSave` with similar props as we do when saving a template as a story - I tried to pass the bare minimum necessary to render the preview. When this new object is returned we pass it to the new serializer `reshapeStoryPreview` which chunks up the new story and adds the few things that aren't part of `getStoryPropsToSave` - much alike what the animation tool preview does to render previews. 
  - After reshaping that we call on edit-story's `getStoryMarkup` which returns markup. 
  - We turn that markup into a string and store it as `previewMarkup` in state.stories context 
  - Now we're back to the `PreviewStory` component! When `previewMarkup` exists we are ready to create our iframe and append it to the view. 
  - When preview is closed the initial useEffect hook cleans up with actions.storyApi.clearStoryPreview to reset the previewMarkup to an empty string.
  - The only other thing that `PreviewStory` is tracking is the available width/height of the screen when the component is loaded and then through the shared `useResizeEffect` hook. 

## To-do

- [x] More karma tests 
- [x] There's something odd happening within the preview after the initial preview is closed and a second is open - the story progresses vertical as it flips pages and sometimes the preview shows blank until you move forward one page? 

## User-facing changes
- Behind a feature flag. When flag is enabled it will allow previews of a template to be rendered when a template card is clicked. 

## Testing Instructions
- See new Views/PreviewScreen in storybook 
<img width="1397" alt="Screen Shot 2020-07-29 at 12 06 50 AM" src="https://user-images.githubusercontent.com/10720454/88768119-77b15580-d12f-11ea-9dfd-38d137f2dbc0.png">

- Set `enableTemplatePreviews` flag in includes/dashboard to true
  -  navigate to explore templates
    -  click on the template preview (outside of either of the existing buttons)
    - See that a preview of that template shows up. 
  - You should be able to close the template with the 'x' button on top right and proceed to open another preview. 
  - You should also be able to open the preview by keying down on 'enter' while focused on a template card. 


---



Addresses #3390
